File: testcodes/uf.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// find
int find(int* node, int sz, int x) {
    if (node[x] != x) 
        node[x] = find(node, sz, node[x]);
    return node[x];
}

// union
void unite(int* node, int sz, int a, int b) {
    int rootA = find(node, sz, a);
    int rootB = find(node, sz, b);
    if (rootA != rootB) 
        node[rootB] = rootA;
}

// connected
int connected(int* node, int sz, int a, int b){
    return find(node, sz, a) == find(node, sz, b);
}

// make node
int* makeNode(int sz) {
    int* arr = malloc(sizeof(int) * sz);
    for (int i = 0; i < sz; ++i) 
        *(arr + i) = i;
    return arr;
}

// print arr. (array, type, size)
void printArr(void* arr, const char* type, int sz) {
    if (strcmp(type, "int") == 0) {
        int *a = arr;
        for (int i = 0; i < sz; ++i)
            printf("%d ", a[i]);
    } 
    else if (strcmp(type, "double") == 0) {
        double *a = arr;
        for (int i = 0; i < sz; ++i)
            printf("%.2f ", a[i]);
    } 
    else if (strcmp(type, "char") == 0) {
        char *a = arr;
        for (int i = 0; i < sz; ++i)
            printf("%c ", a[i]);
    } 
    else {
        printf("[unsupported type]");
    }
    printf("\n");
}


int main() {
    int arrSz = 10;
    char* arr = malloc(sizeof(char) * arrSz);
    // arr의 데이터: A ~ J 의 문자.
    for (int i = 0; i < arrSz; ++i) {
        *(arr + i) = 'A' + i;
    }
    printArr(arr, "char", arrSz);

    // arr의 node 생성.
    int* arrNode = makeNode(arrSz);
    printArr(arrNode, "int", arrSz);

    // node 연결
    unite(arrNode, arrSz, 0, 3);
    unite(arrNode, arrSz, 3, 2);

    // boolean 결과.
    int test1 = connected(arrNode, arrSz, 0, 3);
    int test2 = connected(arrNode, arrSz, 0, 1);
    printf("%d, %d\n", test1, test2);

    // node 직접 print 해보기.
    printArr(arrNode, "int", arrSz);

    free(arr);
    free(arrNode);
    return 0;
}

File: testcodes/UF.java
import java.util.Arrays;

// class my {
//     static <T> void print(T t) { System.out.println(t); }
// }

public class UF {
    private final int size;
    private int[] node;

    public UF(int size) {
        this.size = size;
        this.node = new int[size];
        for (int i = 0; i < this.size; ++i) this.node[i] = i;
    }

    public int find(int n) {
        if (n != node[n]) {
            this.node[n] = find(this.node[n]);
        }
        return this.node[n];
    }

    public void union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);
        if (rootA != rootB) {
            this.node[b] = rootA;
        }
    }

    public boolean isConnect(int a, int b) { return find(a) == find(b); }

    @Override
    public String toString() {
        return Arrays.toString(this.node);
    }

    // main test
    // public static void main(String[] args) {
    //     UF a = new UF(5);
    //     a.union(0, 3);
    //     a.union(3, 2);
    //     a.union(4, 0);
    //     my.print(a.isConnect(0, 2));
    //     my.print(a.isConnect(1, 0));
    //     my.print(a);
    // }
}
